substrate/frame/contracts/src/tests/test_xcm.rs

```diff
@@ -0,0 +1,358 @@
+			None,
+			VersionedXcm::V3(message).encode(),
+			DebugInfo::UnsafeDebug,
+			CollectEvents::UnsafeCollect,
+			Determinism::Enforced,
+		)
+		.result
+		.unwrap();
+
+		assert_return_code!(result, RuntimeReturnCode::Success);
+
+		// Check if the funds are subtracted from the account of Alice and added to the account of
+		// Bob.
+		let initial = INITIAL_BALANCE;
+		assert_eq!(parachain::Assets::balance(0, contract_addr), initial);
+		assert_eq!(ParachainBalances::free_balance(BOB), initial + amount);
+	});
+}
+
+#[test]
+fn test_xcm_execute_filtered_call() {
+	MockNet::reset();
+
+	let contract_addr = instantiate_test_contract("xcm_execute");
+
+	ParaA::execute_with(|| {
+		// `remark`  should be rejected, as CallFilter is not allowed by our CallFilter.
+		let call = parachain::RuntimeCall::System(frame_system::Call::remark { remark: vec![] });
+		let message: Xcm<parachain::RuntimeCall> = Xcm(vec![Transact {
+			origin_kind: OriginKind::Native,
+			require_weight_at_most: Weight::MAX,
+			call: call.encode().into(),
+		}]);
+
+		let result = ParachainContracts::bare_call(
+			ALICE,
+			contract_addr.clone(),
+			0,
+			Weight::MAX,
+			None,
+			VersionedXcm::V3(message).encode(),
+			DebugInfo::UnsafeDebug,
+			CollectEvents::UnsafeCollect,
+			Determinism::Enforced,
+		);
+
+		// The debug message should say that the call was filtered.
+		assert_eq!(
+			std::str::from_utf8(&result.debug_message).unwrap(),
+			"call failed with: CallFiltered"
+		);
+
+		// The call should fail, with an `OutOfGas` error, as the failed xcm::execute will not
+		// refund the max_weight gas passed to it.
+		assert_err!(result.result, crate::Error::<parachain::Runtime>::OutOfGas);
+	});
+}
+
+#[test]
+fn test_xcm_execute_reentrant_call() {
+	MockNet::reset();
+
+	let contract_addr = instantiate_test_contract("xcm_execute");
+
+	ParaA::execute_with(|| {
+		let transact_call = parachain::RuntimeCall::Contracts(crate::Call::call {
+			dest: contract_addr.clone(),
+			gas_limit: 1_000_000.into(),
+			storage_deposit_limit: None,
+			data: vec![],
+			value: 0u128,
+		});
+
+		// The XCM used to transfer funds to Bob.
+		let message: Xcm<parachain::RuntimeCall> = Xcm(vec![
+			Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: 1_000_000_000.into(),
+				call: transact_call.encode().into(),
+			},
+			ExpectTransactStatus(MaybeErrorCode::Success),
+		]);
+
+		ParachainContracts::bare_call(
+			ALICE,
+			contract_addr.clone(),
+			0,
+			Weight::MAX,
+			None,
+			VersionedXcm::V3(message).encode(),
+			DebugInfo::UnsafeDebug,
+			CollectEvents::UnsafeCollect,
+			Determinism::Enforced,
+		);
+
+		// assert that the Transact failed
+		// use crate::tests::{test_xcm::Outcome::Incomplete};
+		assert!(parachain::System::events().iter().any(|r| {
+			matches!(
+				r.event,
+				parachain::RuntimeEvent::PolkadotXcm(pallet_xcm::Event::Attempted {
+					outcome: Outcome::Incomplete(_, XcmError::ExpectationFalse)
+				}),
+			)
+		}));
+	});
```

Comment:
we could also add the fund transfer message from the previous test to the inner (re-entrant) call, and then check the balances stay unchanged.

smth like this 

```suggestion
	ParaA::execute_with(|| {
		let amount: u128 = 42 * CENTS;
		// The XCM used to transfer funds to Bob.
		let message: xcm_simulator::Xcm<()> = Xcm(vec![
			WithdrawAsset(vec![(Here, amount).into()].into()),
			DepositAsset {
				assets: All.into(),
				beneficiary: AccountId32 { network: None, id: BOB.clone().into() }.into(),
			},
		]);

		let transact_call = parachain::RuntimeCall::Contracts(crate::Call::call {
			dest: contract_addr.clone(),
			gas_limit: 1_000_000.into(),
			storage_deposit_limit: None,
			data: VersionedXcm::V3(message).encode(),
			value: 0u128,
		});

		// The XCM used to re-enter the contract call
		let message: Xcm<parachain::RuntimeCall> = Xcm(vec![
			Transact {
				origin_kind: OriginKind::Native,
				require_weight_at_most: 1_000_000_000.into(),
				call: transact_call.encode().into(),
			},
			ExpectTransactStatus(MaybeErrorCode::Success),
		]);

		ParachainContracts::bare_call(
			ALICE,
			contract_addr.clone(),
			0,
			Weight::MAX,
			None,
			VersionedXcm::V3(message).encode(),
			DebugInfo::UnsafeDebug,
			CollectEvents::UnsafeCollect,
			Determinism::Enforced,
		);

		// assert that the Transact failed
		assert!(parachain::System::events().iter().any(|r| {
			matches!(
				r.event,
				parachain::RuntimeEvent::PolkadotXcm(pallet_xcm::Event::Attempted {
					outcome: Outcome::Incomplete(_, XcmError::ExpectationFalse)
				}),
			)
		}));

		// Funds should not change hands, as the inner message declined;
		let initial = INITIAL_BALANCE;
		assert_eq!(ParachainBalances::free_balance(BOB), initial);
	});
```